specs/BrowserProcessExited.md

```diff
@@ -0,0 +1,322 @@
+}
+
+// A new version of the WebView2 Runtime is available, our handler gets called.
+// We close our WebView and set a handler to reinitialize it once the browser
+// process is gone, so we get the new version of the WebView2 Runtime.
+void Environment_NewBrowserVersionAvailable(object sender, object e)
+{
+    StringBuilder messageBuilder = new StringBuilder(256);
+    messageBuilder.Append("We detected there is a new version of the WebView2 Runtime installed. ");
+    messageBuilder.Append("Do you want to switch to it now? This will re-create the WebView.");
+    var selection = MessageBox.Show(this, messageBuilder.ToString(), "New WebView2 Runtime detected", MessageBoxButton.YesNo);
+    if (selection == MessageBoxResult.Yes)
+    {
+        // Save URI or other state you want to restore when the WebView is recreated.
+        _uriToRestore = webView.Source;
+        _coreWebView2Environment.BrowserProcessExited += Environment_BrowserProcessExited;
+        // We dispose of the control so the internal WebView objects are released
+        // and the associated browser process exits. If there are any other WebViews
+        // from the same environment configuration, they need to be closed too.
+        webView.Dispose();
+        webView = null;
+    }
+}
+
+void Environment_BrowserProcessExited(object sender, CoreWebView2BrowserProcessExitedEventArgs e)
+{
+    ((CoreWebView2Environment)sender).BrowserProcessExited -= Environment_BrowserProcessExited;
+    ReinitializeWebView();
+}
+
+void ReinitializeWebView()
+{
+    webView = new WebView2();
+
+    // Restore URI and other WebView state/setup.
+    webView.CreationProperties = (CoreWebView2CreationProperties)this.FindResource("EvergreenWebView2CreationProperties");
+    webView.NavigationStarting += WebView_NavigationStarting;
+    webView.NavigationCompleted += WebView_NavigationCompleted;
+
+    Binding urlBinding = new Binding()
+    {
+        Source = webView,
+        Path = new PropertyPath("Source"),
+        Mode = BindingMode.OneWay
+    };
+    url.SetBinding(TextBox.TextProperty, urlBinding);
+
+    MyWindow.MyDockPanel.Children.Add(webView);
+    webView.Source = (_uriToRestore != null) ? _uriToRestore : new Uri("https://www.bing.com");
+    RegisterForNewVersion();
+}
+```
+
+
+# Remarks
+Note this is an event from the `ICoreWebView2Environment3` interface, not the
+`ICoreWebView2`. The difference between this `BrowserProcessExited` event and
+the `CoreWebView2`'s `ProcessFailed` event is that `BrowserProcessExited` is
+raised for any (expected and unexpected) **browser process** (along its
+associated processes) exits, while `ProcessFailed` is raised only for
+**unexpected** browser process exits, or for **render process**
+exits/unresponsiveness. To learn more about the WebView2 Process Model, go to
+[Process model](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/process-model).
```

Comment:
That documentation page has confusing sentences like

> One browser process is specified per user data folder in a user session that serve any WebView2 requesting process that specifies that user data folder. 

How do you "specify" a browser process for your user data folder?

What is the subject of "serve"? Is it "browser process"? "user data folder"? "user session"?

And I can't puzzle out what "requesting process that specifies that user data folder" is trying to say.

I think the first part of the sentence is trying to say

> Within a user session, all the WebView2 controls that use the same user data folder share a browser process.

I can't figure out what the second part of the sentence is trying to say.

Reply:
I think that doc needs some work to better explain how the browser processes and user data folders work. @pagoe-msft

Reply:
@david-risney to open a bug about updating the doc.

Reply:
Opened item in our backlog.